### DISCUSSION TITLE
How do i check if an element exists?

### DISCUSSION THREAD

**Discussion by GrandaddyShmax** [cefsharp/CefSharp#4131] [Q&A]
Hello!
Trying to make an autofill app and for that i need to detect whether the element exists or not
with the regular VisualStudio C# WebBrowser, I was able to achieve it with this:
`if (webBrowser1.Document.GetElementById(input) != null)`

But having hard time to figure out how to achieve the same effect with CefSharp
I tried this code, but it only makes my app crash:

           `string EvaluateJavaScriptResult;
            var frame = webBrowser1.GetMainFrame();
            foreach (string input in userInput)
            {
                var task = frame.EvaluateScriptAsync("(function() { return document.getElementById('" + input + "'); })();", null);
                task.ContinueWith(t =>
                {
                    if (!t.IsFaulted)
                    {
                        var response = t.Result;
                        EvaluateJavaScriptResult = response.Success ? (response.Result.ToString() ?? "null") : response.Message;
                        userFlag = true;
                    }
                }, TaskScheduler.FromCurrentSynchronizationContext());
            }
`

**Comment by amaitland** [comment]
> There are limits to the complexity of the object graph that can be returned (graphs with cyclic references aren't supported currently), in these cases you may need to turn the JavaScript object into a JSON string with the JavaScript JSON.stringify() method and return that string to your .NET code. Then you can decode that string into a .NET object with something like JSON.NET. See [MDN JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) for more details. See https://stackoverflow.com/a/46881092/4583726 for some guidence on using JSON.stringify with HTMLElement.

As per https://github.com/cefsharp/CefSharp/wiki/General-Usage#2-how-do-you-call-a-javascript-method-that-returns-a-result

It's not possible to directly return a `HTMLElement` via EvaluateScriptAsync as it's too complex.

You can easily move the null check into your `javascript` function so it returns a bool.

The alternative is to use [CefSharp.Puppeteer](https://github.com/cefsharp/Puppeteer#dom-access) which provides an `API` accessing the DOM.
